src/id_allocator.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::result;
+
+/// Error type for IdAllocator usage.
+#[derive(Debug, PartialEq)]
+pub enum Error {
```

Comment:
We might want to add an impl for `std::error::Error`. This costs us nothing (the impl will be empty), but allows the client code use more flexible error handling with boxed errors if needed.